src/Nethermind/Nethermind.Blockchain/Receipts/PersistentReceiptStorage.cs

```diff
@@ -46,18 +48,58 @@ public class PersistentReceiptStorage : IReceiptStorage
             _receiptsRecovery = receiptsRecovery ?? throw new ArgumentNullException(nameof(receiptsRecovery));
             _blocksDb = _database.GetColumnDb(ReceiptsColumns.Blocks);
             _transactionDb = _database.GetColumnDb(ReceiptsColumns.Transactions);
-            _blockFinder = blockFinder ?? throw new ArgumentNullException(nameof(blockFinder));
+            _blockTree = blockTree ?? throw new ArgumentNullException(nameof(blockTree));
             _storageDecoder = storageDecoder ?? ReceiptArrayStorageDecoder.Instance;
+            _receiptConfig = receiptConfig ?? throw new ArgumentNullException(nameof(receiptConfig));
 
             byte[] lowestBytes = _database.Get(Keccak.Zero);
             _lowestInsertedReceiptBlock = lowestBytes is null ? (long?)null : new RlpStream(lowestBytes).DecodeLong();
             _migratedBlockNumber = Get(MigrationBlockNumberKey, long.MaxValue);
+
+            _blockTree.BlockAddedToMain += BlockTreeOnBlockAddedToMain;
+        }
+
+        private void BlockTreeOnBlockAddedToMain(object? sender, BlockReplacementEventArgs e)
+        {
+            if (e.PreviousBlock != null)
+            {
+                RemoveBlockTx(e.PreviousBlock);
+            }
+
+            // Dont block main loop
+            Task.Run(() =>
+            {
+                Block newMain = e.Block;
+
+                // Delete old tx index
+                if (_receiptConfig.TxLookupLimit > 0)
+                {
+                    Block newOldTx = _blockTree.FindBlock(newMain.Number - _receiptConfig.TxLookupLimit.Value);
+                    if (newOldTx != null)
+                    {
+                        ClearTxIndexForBlock(newOldTx);
+                    }
+                }
+            });
+        }
+
+        private void ClearTxIndexForBlock(Block block)
+        {
+            foreach (Transaction transaction in block.Transactions)
+            {
+                _transactionDb[transaction.Hash.Bytes] = null;
+            }
         }
 
         public Keccak FindBlockHash(Keccak txHash)
         {
             var blockHashData = _transactionDb.Get(txHash);
-            return blockHashData is null ? FindReceiptObsolete(txHash)?.BlockHash : new Keccak(blockHashData);
+            if (blockHashData is null) return FindReceiptObsolete(txHash)?.BlockHash;
+
+            if (blockHashData.Length == Keccak.Size) return new Keccak(blockHashData);
+
+            long blockNum = new RlpStream(blockHashData).DecodeLong();
+            return _blockTree.FindBlockHash(blockNum);
```

Comment:
One thing here, in a case of the reorg this block might not have this transaction.
We should change this method to FindBlockNumber? And then check it later? But that would be somewhat incompatible with old layout.

Reply:
Ah probably not an issue as we delete reorged tx mappings, still would be good to have a failsafe later if we don't find the transaction in the block?

Reply:
ok